src/terminator/mod.rs

```diff
@@ -567,17 +569,91 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                     let name = self.memory.read_c_str(name_ptr)?;
                     info!("ignored env var request for `{:?}`", ::std::str::from_utf8(name));
                 }
-                self.write_value(Value::ByVal(PrimVal::Bytes(0)), dest, dest_ty)?;
+                self.write_primval(dest, PrimVal::Bytes(0), dest_ty)?;
+            }
+            
+            "write" => {
+                let fd = self.value_to_primval(args[0], usize)?.to_u64()?;
+                let buf = args[1].read_ptr(&self.memory)?;
+                let n = self.value_to_primval(args[2], usize)?.to_u64()?;
+                trace!("Called write({:?}, {:?}, {:?})", fd, buf, n);
+                let result = if fd == 1 || fd == 2 { // stdout/stderr
+                    use std::io::{self, Write};
+                
+                    let buf_cont = self.memory.read_bytes(buf, n)?;
+                    let res = if fd == 1 { io::stdout().write(buf_cont) } else { io::stderr().write(buf_cont) };
+                    match res { Ok(n) => n as isize, Err(_) => -1 }
+                } else {
+                    info!("Ignored output to FD {}", fd);
+                    n as isize // pretend it all went well
+                }; // now result is the value we return back to the program
+                self.write_primval(dest, PrimVal::Bytes(result as u128), dest_ty)?;
             }
 
             // unix panic code inside libstd will read the return value of this function
             "pthread_rwlock_rdlock" => {
                 self.write_primval(dest, PrimVal::Bytes(0), dest_ty)?;
             }
 
+            // Hook pthread calls that go to the thread-local storage memory subsystem
+            "pthread_key_create" => {
+                let key_ptr = args[0].read_ptr(&self.memory)?;
+                
+                // Extract the function type out of the signature (that seems easier than constructing it ourselves...)
+                // FIXME: Or should we instead construct the type we expect it to have?
+                let dtor_fn_ty = match self.operand_ty(&arg_operands[1]) {
```

Comment:
Fixed.

What about the FIXME? Right now I am extracting the type from the fn that was called. That's slightly odd though: The type is `Option<fn ...>`, but if I try to load from memory at that type, it fails. I have to load at the `fn` type. I suppose this is related to the representation optimization on `Option<Pointer>` -- I thought that would be done "below" MIR level, but it seems like that's not the case and MIR can already observe this optimization?

Anyway, so right now I am extracting the `fn` out of the substitution attached to the `Option`. That doesn't seem very nice, but `mk_fn_ptr` looks rather ugly to call ('d have to somehow create a `PolyFnSig`).